lint_rules/src/main/java/com/hannesdorfmann/adapterdelegates4/lintrules/valrules/ValPatternHandler.kt

```diff
@@ -0,0 +1,43 @@
+package com.hannesdorfmann.adapterdelegates4.lintrules.valrules
+
+import com.android.tools.lint.client.api.UElementHandler
+import com.android.tools.lint.detector.api.JavaContext
+import org.jetbrains.uast.*
+
+/**
+ * @author Paul Karpukhin
+ */
+
+class ValPatternHandler(private val context: JavaContext) : UElementHandler() {
+    private lateinit var node: UVariable
+    private var isAdapterDelegateLibrary = false
+    private var isVal = false
+
+    override fun visitVariable(node: UVariable) {
+        this.node = node
+        checkVal()
+        checkAdapterDelegateLibrary()
+        if (isVal && isAdapterDelegateLibrary) {
+            reportMessage()
+        }
+    }
+
+    private fun checkVal() {
+        isVal = node.isFinal && node.isStatic
```

Comment:
oh actually you could also mess up things if you use the adapterDelegateDSL inside a `companion object`. Thus checking only for `isStatic` makes sense (not isFinal && isStatic)  makes sense  ... 

We could also leave it as it is now and add a second rule that checks for static functions in general to cover the  `companion object` case.

What do you think?

Reply:
It seems that custom lint rules doesn't make sense and there're no memory leaks.

I've tried to use top level val instead of fun for adapter delegate in the project that I'm working on. But there are no memory leaks.  LeakCanary keeps silence about leaking and there is no extraordinary memory cunsumption in Android Profiler.
I even tried to put val and fun in `companion object` but result is the same.

Please take a look at this question https://github.com/sockeqwe/AdapterDelegates/issues/69 from Paul Woitaschek in comments